lesson07.blog-api-server/server.js

```diff
@@ -1,37 +1,213 @@
-contentTypes.set('js', 'text/javascript');
-contentTypes.set('css', 'text/css');
-contentTypes.set('json', 'application/json');
 
+contentTypes.set(Extensions.html, 'text/html');
+contentTypes.set(Extensions.js, 'text/javascript');
+contentTypes.set(Extensions.css, 'text/css');
+contentTypes.set(Extensions.json, 'application/json');
+contentTypes.set(Extensions.txt, 'text/plain');
+contentTypes.set('', 'text/plain');  // default for files with no extension
 
-http.createServer(function (req, res) {
-    const reqUrl = url.parse(req.url);
-    const ext = reqUrl.pathname.split('.')[1];
-    const fileName = reqUrl.pathname.substr(1);
-    const cType = contentTypes.get(ext);
+cacheFiles()
+    .then(() =>
+        http.createServer(handleRequest).listen(PORT, () => {
+            console.log(`Client is available at http://localhost:${PORT}`);
+        }));
 
+// Cache all files in public folder
+async function cacheFiles() {
+    const files = await fs.readdir(PUBLIC_FOLDER);
+    const readPromises = [];
+    files.forEach(file =>
+        readPromises.push(fs.readFile(`${PUBLIC_FOLDER}/${file}`)
+            .then(data => dataCache.set(file, data))
+            .catch(err => err)  // prevent breaking on first possible rejection
+        ));
+    await Promise.all(readPromises);
+    trySetAllDataObject();
+}
 
-    fs.readFile('public/' + fileName, function(err, data) {
-        if(err) {
-            if(err.code == 'ENOENT'){
-                res.writeHead(404, {'Content-Type': 'text/plain'});
-                res.write('Resource no found');
-            }
-            else {
-                res.writeHead(500, {'Content-Type': 'text/plain'});
-                res.write('Server Error');
+function handleRequest(req, res) {
+    switch (req.method.toUpperCase()) {
+        case 'OPTIONS':
+            res.writeHead(200);
+            res.end();
+            return;
+        case 'GET':
+            handleGetRequest(req, res);
+            break;
+        case 'POST':
+            // POST /posts
+            // TODO: handlePostRequest(req, res);
+            break;
+        case 'PUT':
+            // PUT /posts/1
+            // TODO: handlePutRequest(req, res);
+            break;
+        case 'DELETE':
+            // DELETE /posts/1
+            // TODO: handleDeleteRequest(req, res);
+            break;
+        case 'PATCH':
+            // PATCH /posts/1
+            // TODO: handlePatchRequest(req, res);
+            break;
+        default:
+            break;
+    }
+}
+
+function handleGetRequest(req, res) {
+    const { fileName, query } = parseRequest(req);
+
+    if (dataCache.has(fileName)) {
+        const contentType = getContentType(fileName);
+        res.writeHead(200, buildHeader(contentType));
+        res.end(filterData(dataCache.get(fileName), contentType, query));
+    } else {
+        handleNewFileRequest(res, fileName || DEFAULT_RESOURCE, query);
+    }
+}
+
+function parseRequest(req) {
```

Comment:
long function with multiple `return`..
This should be a sign you should split it to smaller functions